app/models/user_data/training_modules_users.rb

```diff
@@ -29,8 +29,9 @@ def furthest_slide?(slide_slug)
     training_progress_manager.slide_further_than_previous?(slide_slug, last_slide_completed)
   end
 
-  def mark_completion(value=true)
-    flags[:marked_complete] = value
+  def mark_completion(value=true, course_id=nil)
+    flags[:marked_complete] = value unless training_module.exercise?
+    flags[course_id] = { marked_complete: value }
```

Comment:
So, does this mean in the event of a training module (i.e. `training_module.training?`) we'll end up with both a key for `:marked_complete` and a key that represents the course id? Won't this end up being duplicate information?

Reply:
We don't use `mark_completion` for anything but exercise modules anyway, do we?

I had noticed this before but then forgot to mention it... as far as I can tell, the line 33 code could just be left out entirely.

Reply:
If that's what we're doing, then we'll want to change the code I quoted above and will need to write some code that updates our existing training modules.

Reply:
@bwreid 
Thanks for pointing out, I'll change the code quoted above.

Regarding updating the existing training modules:
The working of the training modules ( training_module.training? ) stays as it was earlier, but the working of exercises( training_module.exercise? ) changes. The working of the existing exercises will be determined by the earlier hash values <code>flags_hash[:marked_complete]</code> but if the users now click on the mark complete or mark incomplete button, then for those exercises the new nested course hash is created, and from then on, their working will be determined by <code>flags_hash[course_id][:marked_complete]</code>. 
I am not able understand what to update in the existing training modules?

Reply:
@ragesoss Then I'll go forward will removing that line

Reply:
@bwreid @ragesoss Please have a look at the changes.